metals/src/main/scala/scala/meta/internal/metals/MutableMd5Fingerprints.scala

```diff
@@ -14,28 +15,42 @@ final class MutableMd5Fingerprints extends Md5Fingerprints {
   private case class Fingerprint(text: String, md5: String)
   private val fingerprints =
     new ConcurrentHashMap[AbsolutePath, ConcurrentLinkedQueue[Fingerprint]]()
-  def add(path: AbsolutePath, text: String): Unit = {
-    val md5 = MD5.compute(text)
+  def add(
+      path: AbsolutePath,
+      text: String,
+      md5: Option[String] = None
+  ): Unit = {
     val value = fingerprints.computeIfAbsent(
       path,
       { _ =>
         new ConcurrentLinkedQueue()
       }
     )
-    value.add(Fingerprint(text, md5))
+    value.add(Fingerprint(text, md5.getOrElse(MD5.compute(text))))
   }
 
   override def lookupText(path: AbsolutePath, md5: String): Option[String] = {
-    for {
+    val currentLookup = for {
       prints <- Option(fingerprints.get(path))
       fingerprint <- prints.asScala.find(_.md5 == md5)
     } yield {
       // clear older fingerprints that no longer correspond to the semanticDB hash
       prints.clear()
       prints.add(fingerprint)
       fingerprint.text
+    }
 
+    currentLookup match {
+      case None =>
+        val text = FileIO.slurp(path, StandardCharsets.UTF_8)
+        val currentMD5 = MD5.compute(text)
+        if (md5 == currentMD5) {
+          add(path, text, Some(md5))
+          Some(text)
+        } else None
+      case found => found
```

Comment:
What do you think about `.orElse { ... }` instead of pattern matching?

Reply:
Done!